cxl/lib/libcxl.c

```diff
@@ -14541,3 +14541,447 @@ CXL_EXPORT int cxl_memdev_ddr_page_select_get(struct cxl_memdev *memdev)
+                cxl_memdev_get_devname(memdev), cmd->send_cmd->id,
+                CXL_MEM_COMMAND_ID_DDR_HPPR_GET);
+        return -EINVAL;
+    }
+
+    handle_get_ddr_hppr_out = (struct cxl_mbox_handle_get_ddr_hppr_out *)cmd->send_cmd->out.payload;
+    fprintf(stdout, "DDR[0] HPPR is %s\n", (handle_get_ddr_hppr_out->ddr_hppr.hppr_enable[0] == 1)?"Enabled":"Disabled");
+    fprintf(stdout, "DDR[1] HPPR is %s\n", (handle_get_ddr_hppr_out->ddr_hppr.hppr_enable[1] == 1)?"Enabled":"Disabled");
+
+out:
+    cxl_cmd_unref(cmd);
+    return rc;
+}
+
+/* DDR HPPR address info SET */
+#define CXL_MEM_COMMAND_ID_DDR_HPPR_ADDR_INFO_SET CXL_MEM_COMMAND_ID_RAW
+#define CXL_MEM_COMMAND_ID_DDR_HPPR_ADDR_INFO_SET_OPCODE 0xFB2E
+#define CXL_MEM_COMMAND_ID_DDR_HPPR_ADDR_INFO_SET_PAYLOAD_IN_SIZE 8
+
+struct _ddr_addr_info_in {
+  uint8_t ddr_id;
+  uint8_t chip_select; /* 2bit chip select info of faulty row*/
+  uint8_t bank; /* 2bits bank info*/
+  uint8_t bank_group; /* 2bit bank group info */
+  uint32_t row; /* faulty row address */
+}  __attribute__ ((packed));
+
+struct cxl_mbox_handle_set_ddr_hppr_addr_info_in {
+  struct _ddr_addr_info_in hppr_addr_info;
+} __attribute__ ((packed));
+
+CXL_EXPORT int cxl_memdev_ddr_hppr_addr_info_set(struct cxl_memdev *memdev, u8 ddr_id, u8 chip_select, u8 bank_group, u8 bank, u32 row)
+{
+    struct cxl_cmd *cmd;
+    struct cxl_mem_query_commands *query;
+    struct cxl_command_info *cinfo;
+    struct cxl_mbox_handle_set_ddr_hppr_addr_info_in *handle_ddr_hppr_addr_info_set_in;
+    int rc = 0;
+
+    cmd = cxl_cmd_new_raw(memdev, CXL_MEM_COMMAND_ID_DDR_HPPR_ADDR_INFO_SET_OPCODE);
+    if (!cmd) {
+        fprintf(stderr, "%s: cxl_cmd_new_raw returned Null output\n",
+                cxl_memdev_get_devname(memdev));
+        return -ENOMEM;
+    }
+
+    query = cmd->query_cmd;
+    cinfo = &query->commands[cmd->query_idx];
+
+    /* update payload size */
+    cinfo->size_in = CXL_MEM_COMMAND_ID_DDR_HPPR_ADDR_INFO_SET_PAYLOAD_IN_SIZE;
+    if (cinfo->size_in > 0) {
+        cmd->input_payload = calloc(1, cinfo->size_in);
+        if (!cmd->input_payload)
+            return -ENOMEM;
+        cmd->send_cmd->in.payload = (u64)cmd->input_payload;
+        cmd->send_cmd->in.size = cinfo->size_in;
+    }
+
+    handle_ddr_hppr_addr_info_set_in = (void *) cmd->send_cmd->in.payload;
+
+    handle_ddr_hppr_addr_info_set_in->hppr_addr_info.ddr_id = ddr_id;
+    handle_ddr_hppr_addr_info_set_in->hppr_addr_info.chip_select = chip_select;
+    handle_ddr_hppr_addr_info_set_in->hppr_addr_info.bank_group = bank_group;
+    handle_ddr_hppr_addr_info_set_in->hppr_addr_info.bank = bank;
+    handle_ddr_hppr_addr_info_set_in->hppr_addr_info.row = row;
+
+    rc = cxl_cmd_submit(cmd);
+    if (rc < 0) {
+        fprintf(stderr, "%s: cmd submission failed: %d (%s)\n",
+                cxl_memdev_get_devname(memdev), rc, strerror(-rc));
+        goto out;
+    }
+
+    rc = cxl_cmd_get_mbox_status(cmd);
```

Comment:
Need to ensure, if error codes are properly conveyed in case of failure, as PPR on same address is rejected by the device while setting address info.

Reply:
Fixed on the device side